bin/ch-completion.bash

```diff
@@ -0,0 +1,403 @@
+        # generate a completion). I didn't like that, hence the “if”.
+        if [[ -n "$cur" ]]; then
+            compopt -o nospace
+            COMPREPLY=( $(compgen -d -S / -- "$cur") )
+        fi
+        return 0
+        ;;
+    esac
+
+    case "$sub_cmd" in
+    build )
+        # FIXME: opts with args that I need to revisit for this logic
+        #
+        #   * --build-arg KEY[=VALUE]
+        #   * --force-cmd=CMD,ARG1[,ARG2...]
+        #
+        case "$prev" in
+        # Go through a list of potential subcommand-specific opts to see if
+        # “$cur” should be an argument. Otherwise, default to CONTEXT or any
+        # valid option (common or subcommand-specific).
+        -f|--file )
+            compopt -o nospace
+            COMPREPLY=( $(_compgen_filepaths "$cur") )
+            return 0
+            ;;
+        -t )
+            # We can't autocomplete a tag, so we're not even gonna allow
+            # autocomplete after this option.
+            COMPREPLY=()
+            return 0
+            ;;
+        *)
+            # Autocomplete to context directory, common opt, or build-specific opt
+            # --force can take “fakeroot” or “seccomp” as an argument, or no
+            # argument at all. To account for this, we add those two arguments
+            # to the list of compgen suggestions, which allows compgen to
+            # autocomplete to “fakeroot,” “seccomp,” or anything that could
+            # logically follow “--force” with no argument.
+            if [[ "$prev" == "--force" ]]; then
+                extras+="$extras fakeroot seccomp"
+            fi
+            COMPREPLY=( $(compgen -W "$_image_build_opts $extras"  -- "$cur") )
+            # By default, “complete” adds a space after each completed word.
+            # This is incredibly inconvenient when completing directories and
+            # filepaths, so we enable the “nospace” option. We want to make sure
+            # that this option is only enabled if there are valid path
+            # completions for “cur,” otherwise spaces would never be added after
+            # a completed word, which is also inconveninet.
+            if [[ -n "$(compgen -d -S / -- "$cur")" ]]; then
+                compopt -o nospace
+                COMPREPLY+=( $(compgen -d -S / -- "$cur") )
+            fi
+            ;;
+        esac
+        ;;
+    build-cache )
+        COMPREPLY=( $(compgen -W "--reset --gc --tree --dot" -- "$cur") )
+        ;;
+    delete | list )
+        if [[ "$sub_cmd" == "list" ]]; then
+            extras+="$extras -l --long"
+        fi
+        # The following check seems to protects against trying to ls a
+        # non-existent directory, and fixes a bug where the completion function
+        # initialzes an empty storage directory.
+        if [[ -f "$strg_dir" && -n "$(ls "$strg_dir/img")" ]]; then
+            COMPREPLY=( $(compgen -W "$(ls "$strg_dir/img" | sed 's/+/:/g' | sed 's/%/\//g') $extras" -- "$cur") )
```

Comment:
1. https://unix.stackexchange.com/questions/128985
2. Use `sed -E`
3. You can give multiple expressions to one `sed`.
4. Suggest using a delimiter other than `/` to avoid the need for escaping, e.g. `s|%|/|g`.